rust-src/bulletproofs/src/set_non_membership_proof.rs

```diff
@@ -0,0 +1,668 @@
+    transcript.append_message(b"T2", &T_2);
+
+    // Part 4: Evaluate l(.), r(.), and t(.) at challenge point x
+    // get challenge x from transcript
+    let x: C::Scalar = transcript.challenge_scalar::<C, _>(b"x");
+    let mut x_sq = x;
+    x_sq.mul_assign(&x);
+    // Compute l(x) and r(x)
+    let mut lx = Vec::with_capacity(n);
+    let mut rx = Vec::with_capacity(n);
+    for i in 0..n {
+        // l[i] <- l_0[i] + x* l_1[i]
+        let mut lx_i = l_1[i];
+        lx_i.mul_assign(&x);
+        lx_i.add_assign(&l_0[i]);
+        lx.push(lx_i);
+        // r[i] = r_0[i] + x* r_1[i]
+        let mut rx_i = r_1[i];
+        rx_i.mul_assign(&x);
+        rx_i.add_assign(&r_0[i]);
+        rx.push(rx_i);
+    }
+    // Compute t(x)
+    // tx <- t_0 + t_1*x + t_2*x^2
+    let mut tx = t_0;
+    let mut tx_1 = t_1;
+    tx_1.mul_assign(&x);
+    tx.add_assign(&tx_1);
+    let mut tx_2 = t_2;
+    tx_2.mul_assign(&x_sq);
+    tx.add_assign(&tx_2);
+    // Compute the blinding t_x_tilde
+    // t_x_tilde <- z*<1,y^n>*v_rand + t_1_tilde*x + t_2_tilde*x^2
+    let mut tx_tilde = z;
+    tx_tilde.mul_assign(&ip_y_n);
+    tx_tilde.mul_assign(v_rand);
+    let mut tx_s1 = t_1_tilde;
+    tx_s1.mul_assign(&x);
+    tx_tilde.add_assign(&tx_s1);
+    let mut tx_s2 = t_2_tilde;
+    tx_s2.mul_assign(&x_sq);
+    tx_tilde.add_assign(&tx_s2);
+    // Compute blinding e_tilde
+    // e_tilde <- a_tilde + s_tilde * x
+    let mut e_tilde = *s_tilde;
+    e_tilde.mul_assign(&x);
+    e_tilde.add_assign(a_tilde);
+    // append tx, tx_tilde, e_tilde to transcript
+    transcript.append_message(b"tx", &tx);
+    transcript.append_message(b"tx_tilde", &tx_tilde);
+    transcript.append_message(b"e_tilde", &e_tilde);
+
+    // Part 5: Inner product proof for tx = <lx,rx>
+    // get challenge w from transcript
+    let w: C::Scalar = transcript.challenge_scalar::<C, _>(b"w");
+    // get generator q
+    let Q = B.mul_by_scalar(&w);
+    // compute scalars c such that c*H = H', that is H_prime_scalars = (1, y^-1,..,
+    // y^-(n-1))
+    let y_inv = match y.inverse() {
+        Some(inv) => inv,
+        None => return Err(ProverError::DivisionError),
+    };
+    let H_prime_scalars = z_vec(y_inv, 0, n);
+    // compute inner product proof
+    let proof = prove_inner_product_with_scalars(transcript, G, H, &H_prime_scalars, &Q, &lx, &rx);
+
+    // return set membership proof
+    if let Some(ip_proof) = proof {
+        Ok(SetNonMembershipProof {
+            A,
+            S,
+            T_1,
+            T_2,
+            tx,
+            tx_tilde,
+            e_tilde,
+            ip_proof,
+        })
+    } else {
+        Err(ProverError::InnerProductProofFailure)
+    }
+}
+
+/// Error messages detailing why proof verification failed
+#[derive(Debug, PartialEq, Eq)]
+pub enum VerificationError {
+    /// The length of `gens` was less than `|the_set|`
+    NotEnoughGenerators,
+    /// The consistency check for `t_0` failed
+    InconsistentT0,
```

Comment:
Can this be made more understandable? Is it possible to state in a sentence or two what T0 is?